### ISSUE TITLE
Manufacturer Class

### ISSUE THREAD

**Issue by gburton** [gburton/Responsive-osCommerce#720]
`	class manufacturer {
		var $_data = array();

		function __construct($mid) {
      $this->buildManufacturer($mid);
		}

    function buildManufacturer($id) {
			global $languages_id;

			$manufacturer_query = tep_db_query("select m.*, mi.* from manufacturers m, manufacturers_info mi where m.manufacturers_id = '" . (int)$id . "' and m.manufacturers_id = mi.manufacturers_id and mi.languages_id = '" . (int)$languages_id . "'");

			if ( tep_db_num_rows($manufacturer_query) === 1 ) {
        $manufacturer = tep_db_fetch_array($manufacturer_query);

				$this->_data = $manufacturer;
			}
		}

		function getData($key = null) {
      return ( isset($this->_data[$key]) ? $this->_data[$key] : $this->_data );
		}
	}
`

Quick abd easy Manufacturer class.  Any spot any potential problems ?

**Comment by osc2nuke** [comment]
It is missing (the build in oscommerce) cache.

Also this class is very open.
It could use getter methods like:
-getManuId
-getManuName
-getManuUrl
-getManuImage

i would also change the constructor:
`function __construct($mid = NULL) {`
if $mid is NULL ........... get all manufacturers.

Just thought sharing here.
But yeah, it is a good start.

**Comment by gburton** [comment]
getData($key = null)

thus

getData('manufacturers_name') 
is broadly equivalent to a get method 
getManuName

I think.  I can't see a good reason to add in lots of getter methods...by using just one, let us say that (eg) shopowner added in "manufacturers_address" into the manufacturers table...then;

getData('manufacturers_address') is ready to use with no extra changes needed (eg adding in a getter method to the class file).

Thoughts ?

And, btw, good to see you.  U ok now?

**Comment by osc2nuke** [comment]
I'm ok... THANKS BUDDY!!

But in the class provided the method :
`getData` is not used or functional?

As the query is only get data via ID ($mid)

In old V3 there is a good working class:
https://github.com/osCommerce/oscommerce/blob/master/osCommerce/OM/Core/Site/Shop/Manufacturer.php

Just not ( as i suggested ) to get all manufacturers.

But still, i kinda think you see it wrong.
The manufacturer comes by default with osc..
The core collection ( read attributes ) stay in place.
Osc not gonna change that.
So whenever some guy make an addon that wants to to offer an manufacturer address, he just "simply" extend the class.
Keep in mind for THAT the template (front AND back-end needs changes).
THAT is exactly i tried to solve in my previous CRUD/BREAD idea.
And it is possible....... it is a proven concept.

**Comment by gburton** [comment]
$brand = new manufacturer(5);  

echo $brand->getData('manufacturers_name');  // Logitech
echo $brand->getData('manufacturers_image');  // manufacturer_logitech.gif

and so on.

print_r($brand);  // gives array of _data:

Array
(
    [manufacturers_id] => 5
    [manufacturers_name] => Logitech
    [manufacturers_image] => manufacturer_logitech.gif
    [date_added] => 2019-06-04 11:17:21
    [last_modified] => 
    [languages_id] => 1
    [manufacturers_url] => http://www.logitech.com
    [url_clicked] => 0
    [date_last_click] => 
    [manufacturers_description] => 
    [manufacturers_seo_description] => 
    [manufacturers_seo_keywords] => 
    [manufacturers_seo_title] => 
)

See where I'm coming from ?

**Comment by osc2nuke** [comment]
The idea is simple............
There is a EXTRA DB table that checks if you lookup a Manufacturers Image.
So when the system figured........ "hey.......... this table column i MUST threat it like an image" 
We go call tep_image.
If it is a TEXT or ANY other form field, we just KNOW what to output.
And this can be automated.
The template by itself already know what to output and how.
Yeah.... we need to figure how we make modules that have a setting like 2 up to 12 bootstrap positions.
But for me these are side issues.
The overal control is in my opinion more important.
The system will always fallback on default bootstrap output.
If the user use a specific template............ it is up to him or his designer to change it according his running design.
But it will make it so much easier and lighter to adapt oscommerce..

**Comment by osc2nuke** [comment]
Ok.......... now a contribution maker add manufacturer address.
how he hook into the "Manufacturer" $data ?

**Comment by gburton** [comment]
> Ok.......... now a contribution maker add manufacturer address.
> how he hook into the "Manufacturer" $data ?

add in a db table, update it.  

This class will only be for (easy) output.  The insertion of the address,,, that is where YOUR bread comes into play, no?

**Comment by osc2nuke** [comment]
EXACTLY!!!!!!!!

**Comment by gburton** [comment]
Yes.  So this class (or something alike it) can be for the shop-side output and your breadBuilder can be for the admin side input.  Is it (Breadbuilder) today use-able for a average shopowner?

**Comment by osc2nuke** [comment]
> Yes. So this class (or something alike it) can be for the shop-side output and your breadBuilder can be for the admin side input. Is it (Breadbuilder) today use-able for a average shopowner?

Yes it is, oscommerce has a huge DIY audience, what is capable to learn quickly.
Do you remember when "modules" where introduced and how quick the audience adopt it.
Same for Hooks........
If have a blunt user who wanna use oscommerce......... and you get tired to explain them each time.
The point is to make it so easy, and much more automated........... can even explain it IN the admin.
So much resources and keep talking/repeated again and again.
Not healthy for dev's.

**Comment by osc2nuke** [comment]
If H would move the Laravel (first E-commerce system that choose a PHP framework).......... dude!!!!!!!!!!!!!
( i am working on it FYI)

**Comment by osc2nuke** [comment]
https://myoscommerce.com/admin/diy.php

Tab: inject SQL (for example: manufacturer address)
Tab: Form RULE (extend default form)

and it can undo changes
Button: UNDO

So much possibility

**Comment by osc2nuke** [comment]
But let's forget the bread idea.
The array generation as $data is as proven usable.
Kind of what i Sayed. (product class.............. remember?)
Good you pick it up.

**Comment by gburton** [comment]
Let's start off by walking...this class (or something alike it) is a good start.

We replace the admin/manufacturers.php page to make it easier for the shopowner (eg when they add in a new DB table).   perhaps that can be the first "module" of your diy.php idea ?

**Comment by osc2nuke** [comment]
> We replace the admin/manufacturers.php page to make it easier for the shopowner (eg when they add in a new DB table). perhaps that can be the first "module" of your diy.php idea ?

I gonna be honest about this specific problem.
I adopted the concept from a laravel package.
In laravel is NOT used a DB design for multilnguage like is done in oscommerce via language_id (READ: oscommerce is not even consistent as it also use languages_id .... look the s).
So that is already a slight problem.
BUT oscommerce use or have also country code, and that is what Laravel uses.

And it gives option to the site owner to use URL like:
https://mysite.com/NL/product_info.php?products_id=21
BUT, osc not cover so called SLUG ( it does in osc V3) for a product where the admin can give an absolute URL to the product:
https://mysite.com/product/Samsung-Galaxy-Tab

What laravel does: it use all field in one table.
So product and product_description is in one table.
YET if switch to a non-default language, it go look to a table called "translations"
And there we have :

`-- Tabelstructure for tabel `translations`
--

CREATE TABLE `translations` (
  `id` int(10) UNSIGNED NOT NULL,
  `table_name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `column_name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `foreign_key` int(10) UNSIGNED NOT NULL,
  `locale` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `value` text COLLATE utf8mb4_unicode_ci NOT NULL,
  `created_at` timestamp NULL DEFAULT NULL,
  `updated_at` timestamp NULL DEFAULT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;`

The system knows where to look.

It is a consistent approach.
Yet oscommerce goes a little wrong in:

products-products_description
categories_categories_description
manufacturers:manufacturers_info

There is no consistency in how the tables are build up.
How that go work for add-on:
blog- blog_content
reviews_reviews_stuff

Like that it just never going to work.
It will be and stay a mess.

**Comment by osc2nuke** [comment]
It is the reason why the BREAD idea fail. or any CRUD appliction for oscommerce.
UNLESS......... we use the other concept i made.
The modulair admin.
So same principe as for the front-end witht he modules and hooks.
But then for the admin.
There we can make "separate" modules for each, yet still use the BREAD stuff.
It is not the logic i would go for 1st place.
But hey......... if can adopt the idea and wiling to make concessions...... i'm in.

**Comment by osc2nuke** [comment]
> $brand = new manufacturer(5);
> 
> echo $brand->getData('manufacturers_name'); // Logitech
> echo $brand->getData('manufacturers_image'); // manufacturer_logitech.gif
> 
> and so on.
> 
> print_r($brand); // gives array of _data:
> 
> Array
> (
> [manufacturers_id] => 5
> [manufacturers_name] => Logitech
> [manufacturers_image] => manufacturer_logitech.gif
> [date_added] => 2019-06-04 11:17:21
> [last_modified] =>
> [languages_id] => 1
> [manufacturers_url] => http://www.logitech.com
> [url_clicked] => 0
> [date_last_click] =>
> [manufacturers_description] =>
> [manufacturers_seo_description] =>
> [manufacturers_seo_keywords] =>
> [manufacturers_seo_title] =>
> )
> 
> See where I'm coming from ?

You aware of this approach made the base of "template-systems"?

**Comment by osc2nuke** [comment]
This is what WP say:
We are WP and you can change us to:
-PIM, 
-ERP, 
-CMS, 
-AMS, 
-MAP, 
-CRM
or ............
We are oscommerce and you can be:
-PIM, 
-ERP, 
-CMS, 
-AMS, 
-MAP, 
-CRM
or a combination.

thoughts?

**Comment by osc2nuke** [comment]
Gary............ everything is a page.
Page A ( type news item)
Page B (type product)
etc.. etc....
only need 1 table called : pages.
all you visit is a page........ right?
Wake up!

**Comment by gburton** [comment]
I'm only interested in making small changes...not big architecture changes.  
There simply isn't the support there (from those using osCommerce) to make it worthwhile.

**Comment by osc2nuke** [comment]
> I'm only interested in making small changes...not big architecture changes.
> There simply isn't the support there (from those using osCommerce) to make it worthwhile.

test it.
release it.
We will see who "moves on".
That is how it is done.

looser mentality: There simply isn't the support there (from those using osCommerce) to make it worthwhile.

winner mentality: There simply is support there (from those using osCommerce) to make it worthwhile.

**Comment by osc2nuke** [comment]
The only thing you say is...............:  I wait for H.
You are NOT his BITCH.

**Comment by gburton** [comment]
I'm not willing to put my time into osCommerce.  5 minutes here and there, that's all.

Don't start this SHIT again Henry.

**Comment by osc2nuke** [comment]
So.......... keep following.
H release makes you learn all new again. (His way, given the fact that all of that is already invented and used all around the world, yet you choose to follow the learning guy.......... ignoring your learning capability, how does that sound.......... it sound like shit to you?
)

**Comment by osc2nuke** [comment]
Do you remember i mentioned "guzzle" ..............
https://github.com/haraldpdl/oscommerce/tree/template
go find the introduction of Guzzle in this branche.

You are to scared to loose your audience.
It is not you who fucks up.
It is H.
Why you so scared........ you are a master.

**Comment by osc2nuke** [comment]
The only skilled coder in oscommerce left is you.
And yeah............ the hell......... i have to say it........ even you are an idiot sometimes.
You will not gain anything from H unless you do it yourself.

**Comment by osc2nuke** [comment]
> Don't start this SHIT again Henry. I held out my hand, don't bite it - I'll simply turn off again for another 6 months.

put your hand in your ass..
i am not just a guy who simply comply.

BTW: as deep as possible!

**Comment by gburton** [comment]
This is not the place for this conversation.  Let it end.

I'll introduce the manufacturers class this month.  I'm now going to remove your last few comments as they add nothing to the class discussion.

**Comment by osc2nuke** [comment]
why not use and alter the current class:
`<?php
/**
 * osCommerce Online Merchant
 * 
 * @copyright Copyright (c) 2011 osCommerce; http://www.oscommerce.com
 * @license BSD License; http://www.oscommerce.com/bsdlicense.txt
 */

  namespace osCommerce\OM\Core\Site\Shop;

  use osCommerce\OM\Core\Registry;

  class Manufacturer {
    protected $_data = array();

    public function __construct($id) {
      $OSCOM_PDO = Registry::get('PDO');

      $Qmanufacturer = $OSCOM_PDO->prepare('select manufacturers_id as id, manufacturers_name as name, manufacturers_image as image from :table_manufacturers where manufacturers_id = :manufacturers_id');
      $Qmanufacturer->bindInt(':manufacturers_id', $id);
      $Qmanufacturer->execute();

      $result = $Qmanufacturer->fetch();

      if ( $result !== false ) {
        $this->_data = $result;
      }
    }

    function getID() {
      if ( isset($this->_data['id']) ) {
        return $this->_data['id'];
      }

      return false;
    }

    function getTitle() {
      if ( isset($this->_data['name']) ) {
        return $this->_data['name'];
      }

      return false;
    }

    function getImage() {
      if ( isset($this->_data['image']) ) {
        return $this->_data['image'];
      }

      return false;
    }
  }
?>`

What is wrong with it?
Curious to know.

**Comment by osc2nuke** [comment]
or a "functional" class that perhaps would work out of the box:
`<?php
	/*
		$Id: account.php 207 2005-09-26 01:29:31 +0200 (Mo, 26 Sep 2005) hpdl $
		
		osCommerce, Open Source E-Commerce Solutions
		http://www.oscommerce.com
		
		Copyright (c) 2005 osCommerce
		
		This program is free software; you can redistribute it and/or modify
		it under the terms of the GNU General Public License v2 (1991)
		as published by the Free Software Foundation.
	*/
	
	class osC_Manufacturer {
		var $_data = array();
		
		function osC_Manufacturer($id) {
			global $osC_Database;
			
			$Qmanufacturer_Query = tep_db_query("select manufacturers_id as id, manufacturers_name as name, manufacturers_image as image from manufacturers where manufacturers_id = '" . (int)$id . "' ");
			$Qmanufacturer = tep_db_fetch_array($Qmanufacturer_Query);
			
			if ( tep_db_num_rows($Qmanufacturer_Query) === 1 ) {
				$this->_data = $Qmanufacturer;
			}
		}
		
		function getID() {
			if (isset($this->_data['id'])) {
				return $this->_data['id'];
			}
			
			return false;
		}
		
		function getTitle() {
			if (isset($this->_data['name'])) {
				return $this->_data['name'];
			}
			
			return false;
		}
		
		function getImage() {
			if (isset($this->_data['image'])) {
				return $this->_data['image'];
			}
			
			return false;
		}
	}
?>
`

**Comment by osc2nuke** [comment]
Dude......... i did that long time ago:
https://github.com/osc2nuke/osCommerce-Rsp-Variants/blob/master/includes/classes/manufacturer.php

**Comment by osc2nuke** [comment]
You do NOT need t change the whole system or architect a whole new system to "offer" a user an extra field for:
- manufacturer_address
to this class.
Is that open for discussion here?

**Comment by osc2nuke** [comment]
What if we do on the class:
`			$Qmanufacturer = tep_db_fetch_array($Qmanufacturer_Query);
			
			//check fields
			foreach($Qmanufacturer as $field => $type){
				//we call the db table where we told beforehand to lookup each field
				//we instruct the system to handle each field as a form output
				// a text is a text.
				// a varchar is a ??
				// so much options
				//how we put out a adress field?
				// how we store it?
			}`

**Comment by gburton** [comment]
If the topic is a manufacturers class or linked topics, I'm all ears.  
If it is anything else, do it somewhere else...agreed?

Let's take the example of being able to add/edit/delete a manufacturers_address.  
What's the easiest way forward that;

1.  does not need big changes in architecture
2.  is simple enough for average shopowner to diy

And let's take the example of that same manufacturers_address now that it is added to the DB and data exists for it for a given manufacturer.  How does the shopowner output it in a way that;

1.  does not need big changes in architecture
2.  is simple enough for average shopowner to diy

Look at the example class you copy/pasted up there...how does shopowner output manufacturers_address (unless he has the knowledge to extend the class, and let's be 100% on that...there are not many (any) shopowners who can do that).

**Comment by osc2nuke** [comment]
it is impossible if NOT connect the DB field to a desired outcome.
The field ( either core or new field HAS to be determined upfront to tell it what to do).
I try to learn you something different.
This does not mean change architecture.......... just extend it.

**Comment by osc2nuke** [comment]
are you agree we already have in oscommerce the file: html_output.php
and given by that ( regardless it is to the current standard ( what it is NOT...but hey.....))

Are you agree that if use the basic form input/output rules.
With them can do an easy admin setting/configuration and user output.
Do you agree that forms are so much covered in any programming language , so that they can be handled?

**Comment by osc2nuke** [comment]
If you NOT agree.........
i better move on.

**Comment by osc2nuke** [comment]
Look, the osc front-end already foud it's way via a module.
it mean same as for a product,
it has a module for the title, the price , the description and so on.

I talk about the admin.
How introduce a new field............ right?
It is an interesting discussion.

**Comment by osc2nuke** [comment]
Keep in mind, i offered same principe as in the front-end, for the admin.
It means a module based admin.
So yeah........ it could mean we load admin modules.
And yeah.......... we might load a module that just only collect a field.
YET.............. for that to make you understand is you have to adopt or look into:
https://github.com/osc2nuke/oscommerce-2.3.4.1-CE

It might not look that straight forward, but it might can be an option to look into.
The concept allows it.
I think you even can see that.
if it is the nest solution?
But hey.... you did it for the front-end.

**Comment by osc2nuke** [comment]
let us just agree on the point made here.
(the modulair admin have much more other potential............ i think you see that)
But you also must agree, you never cared for the admin.
But now you are asking, so?
I give a good concept/solution.
Broadly based on the system you use for the front-end.
I do not think you should or can ignore.
It can be solid............ discusable, perhaps not even the best what can get........ but i think it can work for what we have to deal with.
Especially the "little thingys"......... 
perhaps we overkill the modular approach.
But let's overkill it right.

**Comment by osc2nuke** [comment]
And i know you like the admin idea.
So if can we make it work for a table field adoption.
Why not?
There must be made anyway a representation for it on the front-end via a module.

**Comment by osc2nuke** [comment]
i wait for your response.
You asked for an alternative to tackle the problem.
Express yourself.

**Comment by osc2nuke** [comment]
perhaps we can make modules that represent an extention:
`  class db_manu_address {
    var $code;
    var $group;
    var $title;
    var $description;
    var $sort_order;
    var $enabled = false;`
    var fields = [];
